dags/xcom_gcs_ds.py

```diff
@@ -105,8 +123,22 @@ def train(df: pd.DataFrame):
         return np.mean(n_scores)
 
     @task
-    def fit(accuracy: float, df: pd.date_range): 
+    def fit(accuracy: float, **kwargs): 
+        """Fit the final model
+        
+        Determines if accuracy meets predefined threshold to go ahead and fit model on full data set.
+
+        Returns lightgbm model as json via XCom to GCS bucket.
+        
+
+        Keyword arguments:
+        accuracy -- average accuracy score as determined by CV. 
+        """
         if accuracy >= .8:
+
+            # Reuse data produced by the feauture_engineering task by pulling from GCS bucket via XCom
+            df = kwargs['ti'].xcom_pull(task_ids='feature_engineering')
```

Comment:
It bothers me a little bit that we're both using the taskflow API as well as doing a xcom_pull. I mean, I don't think the usage is wrong, it just takes away the 'magic' a little.

Reply:
Hm - yeah good catch. @fhoda is there a more natural way to pull this?

Reply:
This is the only way I get the graph look like what we discussed while also reusing data from a task not directly prior to this one. The previous way I had it was more natural to Taskflow and would address the concern here but then take us back to where we started in terms of the rendering of the graph.

Reply:
I assume you all don't want the extra edge between "feature_engineering" and "fit" when passing the out of "feature_engineering" into this function?
![image](https://user-images.githubusercontent.com/48934154/133663677-5568fe3e-eab4-440b-ac87-84a1da77e8e6.png)

Understandable but I agree it's not taking full advantage of the TaskFlow API.  However, you can update the function signature to directly access the `TaskInstance` object without changing how it's called:
```python
def fit(accuracy: float, ti=None): 
    ...

    # Reuse data produced by the feauture_engineering task by pulling from GCS bucket via XCom
    df = ti.xcom_pull(task_ids='feature_engineering')
    ...
```

Reply:
This also _may_ be an option but I'm not sure how this will work when the desired input would be a `Dataframe`. If this approach doesn't work, I'd be skeptical if setting `render_template_as_native_obj` to True at the DAG level would properly render the value as a `Dataframe` since it may not be considered a "native Python type".

```python
@task
    def fit(accuracy: float, feature_dataset: pd.DataFrame, **kwargs):
...
fit(accuracy, feature_dataset="{{ ti.xcom_pull(task_ids='feature_engineering') }}")
```

Reply:
So... since xcom is pulling the data from the same location, couldn't we make both data passages (train->fit and feature_engineering->fit) equally native? What I am suggesting is a change to the taskflow api rather than one to this PR. It should be relatively straightforward to abstract away `ti.xcom_pull(task_ids='feature_engineering')` by updating the `@task` decorator. Eventually we should support pulling featurized data freely from a feature store even if it's not part of the same branch of the DAG (or even the same DAG). WDYT @josh-fell?

Reply:
100% agree. Even the abstraction included with the `.output` property of operators and `XComArgs` doesn't exactly mirror the classic `xcom_pull()` functionality which gives users access to `XComs` across DAG and execution dates plus access to specific items directly in an `XCom`.  An abstraction you are putting forth would be highly beneficial and I'd love to see it implemented.  Perhaps Daniel has something like this baking with his current decorator-wizardry activities.

I was only suggesting some alternatives if we didn't want to show a non-serial pipeline path which is generated when we try to `fit(accuracy, features)`.

Reply:
I have applied @josh-fell 's first suggestion. I am not really opinionated on this matter and can test out the second one if there are strong feelings towards using that syntax.